vulcan/vulcan.go

```
@@ -0,0 +1,162 @@
+	Scannable   bool         `json:"Scannable"`
+	AssetType   AssetType    `json:"AssetType"`
+	Identifier  string       `json:"Identifier"`
+	Annotations []Annotation `json:"Annotations"`
+}
+
+// Team represents the "team" model as defined by the Vulcan async API.
+type Team struct {
+	ID          string `json:"Id"`
+	Name        string `json:"Name"`
+	Description string `json:"Description"`
+	Tag         string `json:"Tag"`
+}
+
+// AssetType represents the "assetType" model as defined by the Vulcan async
+// API.
+type AssetType string
+
+// Annotation represents the "annotation" model as defined by the Vulcan async
+// API.
+type Annotation struct {
+	Key   string `json:"Key"`
+	Value string `json:"Value"`
+}
+
+// AssetMetadata represents the "assetMetadata" model as defined by the Vulcan
+// async API.
+type AssetMetadata struct {
+	Version    string    `json:"Version"`
+	Type       AssetType `json:"Type"`
+	Identifier string    `json:"Identifier"`
+}
+
+// Client is a Vulcan async API client.
+type Client struct {
+	proc stream.Processor
+}
+
+// AssetHandler processes an asset. isNil is true when the value of the stream
+// message is nil.
+type AssetHandler func(id string, payload AssetPayload, isNil bool) error
+
+// NewClient returns a client for the Vulcan async API using the provided
+// stream processor.
+func NewClient(proc stream.Processor) Client {
+	return Client{proc}
+}
+
+// ProcessAssets receives assets from the underlying stream and processes them
+// using the provided handler. This method blocks the calling goroutine until
+// the specified context is cancelled.
+func (c Client) ProcessAssets(ctx context.Context, h AssetHandler) error {
+	return c.proc.Process(ctx, AssetsEntityName, func(msg stream.Message) error {
+		version, typ, identifier, err := parseMetadata(msg)
+		if err != nil {
+			return fmt.Errorf("invalid metadata: %w", err)
+		}
+
+		if !supportedVersion(version) {
+			// Do not process asset on version mismatch.
```

I would maybe write and informational log here just to make easier to observe if there are messages in the topic that are not being processed.

---

Alternatively we could return a concrete error when a message has not been processed because it has an incompatible version, and let the ``run`` function log the error.

---

As we agreed in our last discussion, now an error is returned as soon as an asset with an unsupported version is received.